Allow textures to be encoded/decoded in Linear,sRGB,RGBE,RGBM, etc

bhouston opened the pull request:
This PR implements the idea discussed here a year ago:

https://github.com/mrdoob/three.js/issues/6593

To understand the theory behind this, please read this page:

http://lousodrome.net/blog/light/2013/05/26/gamma-correct-and-hdr-rendering-in-a-32-bits-buffer/

I have created functions for texelEncode and texelDecode.  I have not filled out all the formats in both situations, but I think that is really easy for others to contribute there on a need basis.

Right now I only use the encoding when accessing textures that are expected to be outside of the standard [0-1] range, which are generally light generating textures, either EnvMaps or Emissivemaps.  I applied the encoding sparingly for now because I didn't want to unnecessarily use uniforms (all other maps generaly do not use such encodings) when they may be at a premium on mobile devices.

Clara.io uses this change in its custom ThreeJS in order to support HDR IBL maps without requiring the float texture extension.  I believe SketchFab and Marmoset do similar things as well.  Thus this is an industry standard change.

All existing code works as it did, so this should be a safe merge.  I think it is also something that can easily be improved upon by others on this project as it is a very clean design.

/ping @WestLangley @spidersharma03

bhouston committed:
add ability to have texel encodings (RGBE,RGBM,etc.) on envMap and emissiveMap that can contain unbounded intensities.

bhouston committed:
bug fixes.

bhouston committed:
add links to helpful websites.

MiiBond commented:
Very nice!

Incidentally, I ended up abandoning this approach of decoding these formats in the shader every frame. As you probably know, you can't use bilinear filtering on RGBE or LogLuv images and I found that I really needed mipmapping and filtering for optimal quality for IBL. I've ended up using RGBE but I unpack it into a floating-point texture at load time. Support is in the high 90's for float textures according to WebGL stats with the exception of IE11, which is at 92%. Still, those are pretty good numbers.

bhouston commented:
That makes sense. :)

I'd still advocate pulling this in because it does introduce the ability to encode/decode textures, and if you want you can use these decoders/encoders to write to floating point textures and then access thus via ENCODING_Linear.  So I view most of this code as necessary for pre-decoding as well.

bhouston commented:
I found the artifacts minimal when using RGBM7 and RGBM16 formats though, in part because M interpolates better than E:  https://clara.io/view/d59d26ab-c26e-4fa0-8907-d52c0c8acc1a

BTW we had this discussion a year ago here: https://github.com/mrdoob/three.js/issues/6593  That said I still think we should get this merged in because it enables decoding-based pipelines, whether pre or inline.

bhouston commented:
Also I agree with you that pre-decoding is better, just more work, as I described here: https://github.com/mrdoob/three.js/issues/6383#issuecomment-97844411

MiiBond commented:
Totally agree. I didn't mean to argue against the pull request. Only sharing my experience.

RGBM definitely works better but I was having quality issues with some HDR maps. I think I remember having better luck with RGBD.

bhouston commented:
Heh.  We had the RGBD discussion as well last year: https://github.com/mrdoob/three.js/issues/6593#issuecomment-106413150  Because of that, I added RGBD in this PR but commented out until someone wants to add formal support.

bhouston commented:
I found that RGBM's main limitation was its limited dynamic range.  RGBM7 is limited to [0-7], and RGBM16 just goes from [0-16], which is better than LDR but far from the RGBE range of [0,2^127].

bhouston committed:
use standard GAMMA_FACTOR define.

mrdoob commented on src/Three.js:
```diff
@@ -305,3 +305,13 @@ THREE.WrapAroundEnding = 2402;
+//THREE.LogLuv = 3003; TODO
+THREE.RGBM7 = 3004;
+THREE.RGBM16 = 3005;
+//THREE.RGBD = 3006; TODO
```
We should probably do:

```javascript
THREE.LinearTextureEnconding = 3000;
THREE.sRGBTextureEnconding = 3001; // AKA gamma 2.2.
THREE.RGBETextureEnconding = 3002; // AKA Radiance
THREE.RGBM7TextureEnconding = 3004;
THREE.RGBM16TextureEnconding = 3005;
```

or start doing this:

```javascript
THREE.TEXTURE_ENCONDINGS.LINEAR = 3000;
THREE.TEXTURE_ENCONDINGS.SRGB = 3001; // AKA gamma 2.2.
THREE.TEXTURE_ENCONDINGS.RGBE = 3002; // AKA Radiance
THREE.TEXTURE_ENCONDINGS.RGBM7 = 3004;
THREE.TEXTURE_ENCONDINGS.RGBM16 = 3005;
```

bhouston replied:
I'll make this change now.  Somehow I missed this comment.

bhouston committed:
adopt new naming scheme per @mrdoob's recommendation.

bhouston committed:
add support for Texture.encoding to Material.map

bhouston committed:
set backwards compatible default.

bhouston commented:
Added support for Material.map to have an encoding to deal with @mattdesl's sRGB usecase.

bhouston committed:
make encodings.glsl more reusable.

bhouston committed:
begin on swithcing texture encodings from uniforms to macros.

bhouston committed:
adopt macro-defined encodings, remove uniform defined encodings.

bhouston committed:
add preprocessor that by default uses the ShaderChunk library.

bhouston committed:
remove THREE.DefaultEncoding and also old uniforms for encoding.

bhouston committed:
get macros for encodings working.

bhouston committed:
cleanup PR.

bhouston commented:
@mattdesl I've updated this PR to move away from the uniform-based selection of encodings in the glsl (which used up precision uniforms) to a preprocessor define system that is actually texture-specific rather than material specific.  This will allow us to add as many texture-specific encodings as we want.

I've also written a shader preprocessor so that I can use "#include" statements in my shaders.

bhouston committed:
remove unused encoding functions.

bhouston commented:
@WestLangley @mrdoob @MiiBond @tschw any thoughts on this PR?  I would like to get this PR merged so that @spidersharma03 and myself can get the PMREM generator, which is dependent upon encodings, finished.

I think with the new preprocessor define-based system there is no GLSL run-time cost to this PR's increased flexibility with regards to texture encodings, although the cost to compile is slightly higher.

It also includes a useful but short WebGLShaderPreProcessor that can be further expanded - right now it supports only top-level "#include" statements.

bhouston committed:
simplify preprocessor, no need for reserve if I use pop.

bhouston committed:
Merge branch 'dev' of git@github.com:mrdoob/three.js into texture_encoding

WestLangley commented:
Nice work, @bhouston !

bhouston committed:
fix bug with regards to emissiveMapTexelToLinear...

bhouston committed:
fill in the LogLuv, RGBM, and RGBD decoders/encoders.

bhouston committed:
ensure that particles also use map encodings.

tschw commented:
Linearization / delinearization functions in the [sRGB standard](http://www2.units.it/ipl/students_area/imm2/files/Colore1/sRGB.pdf) actually are

```glsl
float srgbLinearize( float c ) {

    return c <= 0.04045 ? c * 0.0773993808 :
            pow( c * 0.9478672986 + 0.0521327014, 2.4 );

}

float srgbGammaCorrect( float c ) {

    return c <= 0.0031308 ? c * 12.92 : pow( c, 0.41666 ) - 0.055;

}
```

Had the GLSL laying around. I'm not sure whether the difference justifies the extra cost.

bhouston commented:
@tschw, I am for using the sRGB standard decoding/encoding, but it does affect backwards compatibility because right now our gamma is actually 2.0 (probably for optimization purposes), it isn't even the proper near approximate sRGB of 2.2.

Maybe we could try to have the best of both worlds and keep around "GammaToLinear"/"LinearToGamma" that uses a gamma factor of 2.0, and also introduce a sRGBToLinear/LinearTosRGB.  Thus we split it so that one is absolutely correct and the other is equivalent to what we are currently doing?

BTW are you sure it is the full sRGB to Linear transform -- there are extra steps here: https://en.wikipedia.org/wiki/SRGB

tschw commented:
> Thus we split it so that one is absolutely correct and the other is equivalent to what we are currently doing?

Sounds reasonable.

> BTW are you sure it is the full sRGB to Linear transform -- there are extra steps here:
> https://en.wikipedia.org/wiki/SRGB

Pretty much so: The further steps are the RGB <-> XYZ conversion. That color space has a physically unambiguous definition and serves as "ground truth" for the rendering intent. This way, it's perfectly clear what to do in case an sRGB-encoded image is to be displayed on a calibrated display. Often however, sRGB is used "the other way around": To have a reasonable definition when no precise data about spectral characteristics is available, where device vendors are encouraged to comply with it - more or less.
 
Citing https://en.wikipedia.org/wiki/SRGB
> **It is important to note that these linear RGB values are *not* the final result.**

If you just want linear intensities and stick with the (assumed) display RGB primaries **they actually *are*.** You have to stop before reaching the "final result", which refers to XYZ.

https://en.wikipedia.org/wiki/Color_management
https://en.wikipedia.org/wiki/CIE_1931_color_space

bhouston committed:
add true sRGB support for texture.encodings.

bhouston committed:
try hand at using bvec3 with mix for optimizing sRGB linearization.

bhouston commented:
@tschw I'm implemented your suggestion using your provided code - thank you.

I made an attempt at vectorizing the sRGB conversions as well here as I think (?) it should be faster this way: https://github.com/bhouston/three.js/commit/bdcb83ab28b8eecb443d71144ebe13e5d19a03c2

bhouston committed:
fix order bug with mix-bvec3

mattdesl commented on src/renderers/shaders/ShaderChunk/encodings.glsl:
```diff
@@ -0,0 +1,93 @@
+  return vec4( mix( pow( value.rgb * 0.9478672986 + vec3( 0.0521327014 ), vec3( 2.4 ) ), value.rgb * 0.0773993808, lessThanEqual( value.rgb, vec3( 0.04045 ) ) ), value.w );
+}
+vec4 LinearTosRGB( in vec4 value ) {
+  return vec4( mix( pow( value.rgb, vec3( 0.41666 ) ) - vec3( 0.055 ), value.rgb * 12.92, lessThanEqual( value.rgb, vec3( 0.0031308 ) ) ), value.w );
```
Getting compile errors here:

```
gl.getShaderInfoLog() vertex ERROR: 0:162: 'mix' : no matching overloaded function found 
```

Looks like `mix` params are mismatched.

mattdesl replied:
Ok, it seems like `lessThanEqual` returns a `bvec` which maybe can't be passed into `mix`?

tschw replied:
> Ok, it seems like lessThanEqual returns a bvec which maybe can't be passed into mix?

Confirmed. Not when targeting OpenGL ES 2. 

It only looks like it might be available [here](https://www.opengl.org/sdk/docs/man4/html/mix.xhtml). But the overload isn't listed in the [spec](https://www.khronos.org/registry/gles/specs/2.0/GLSL_ES_Specification_1.0.17.pdf) (see 8.3 "Common Functions").

tschw commented:
> I made an attempt at vectorizing the sRGB conversions as well here as I think (?) it should be faster this way

...given that `mix` causes `pow` to be executed conditionally, figuring pipeline availability may be the limiting factor, which is why I was shy to use it. However, I like the vectorized variant, because it's more concise.

The impact of vectorization on performance will probably (and hopefully) vanish as technology advances - it can (as there are implementations) and should be automated at lower levels, since machines don't overlook things and are not limited by having to keep some source code maintainable. Also, executing vectorized code needs more state, thus the only way to actually get it right in general is by knowing the device-dependent number of available registers.

bhouston committed:
cast bvec3 to vec3 so that mix works.

bhouston committed:
Merge branch 'dev' of git@github.com:mrdoob/three.js into texture_encoding

bhouston commented:
@mattdesl @tschw Thanks for the help!  I just added a cast from bvec3 to vec3 and it appears to work now. :)

I had some problems with the test webserver I was using, it was excessively caching three.min.js which obviously caused a bunch of issues.

mattdesl commented:
Another problem just came up with this: It doesn't re-compile the shader when a texture encoding changes (from what I can tell). The situation it happened for me was:

- new CompressedTexture is created from `CompressedTextureLoader`
- The texture is set on a mesh even though it hasn't been loaded (user expects black or something)
- GLSL is inlined first time rendering
- After texture loads, user may set the `encoding` to THREE.sRGBEncoding, but even if `needsUpdate` is true it will not re-compile any shaders associated with that texture

bhouston commented:
@mattdesl I am unsure what causes the shader to recompile at the moment.  I do set the parameter string and it should be unique based on the chosen encodings as you can see I include them:

https://github.com/bhouston/three.js/blob/texture_encoding/src/renderers/webgl/WebGLPrograms.js#L99

And i know it uses that getParameters() call to check the program cache here in initMaterials:

https://github.com/mrdoob/three.js/blob/426bbdaebce8164d78bff8acb07f7f0d2819c3c7/src/renderers/WebGLRenderer.js#L1404

And initMaterials is called when the material.needsUpdate is set:

https://github.com/mrdoob/three.js/blob/426bbdaebce8164d78bff8acb07f7f0d2819c3c7/src/renderers/WebGLRenderer.js#L1609

So I am unsure where it is breaking in this chain.

mattdesl commented:
Sorry – to clarify, I think `material.needsUpdate` will recompile with the right encoding, but changing `material.map.encoding` (with or without `material.map.needsUpdate`) will not propagate up to modify the shader. This is maybe something that can be solved with documentation, though. :smile:

bhouston commented:
@mattdesl No changes to texture.* properties will automaticlaly cause texture.needsUpdate to set.  So this issue is pre-existing and effects all texture.* properties equally.

bhouston commented:
@mattdesl The potential fix, which is outside of the scope of this PR, for your issue, is to convert the fields into properties that have getter/setter functions that check if the value is actually being changed and if it is, it auto-triggers needsUpdate.  There are downsides to such automation though so it is clearly outside of the scope of this PR.

bhouston commented:
@mattdesl In your opinion, is this ready to merge into ThreeJS?

mattdesl commented:
My point was, even if you set map.needsUpdate to true, it will not recompile the shader with new defines. Instead, you need to set material.needsUpdate to true on all materials that use those maps. This is not intuitive. 

However since this whole field is for advanced use I think it's OK as long as we document it somewhere. 

Sent from my iPhone

> On Feb 22, 2016, at 4:28 PM, Ben Houston (Clara.io) <notifications@github.com> wrote:
> 
> @mattdesl In your opinion, is this ready to merge into ThreeJS?
> 
> —
> Reply to this email directly or view it on GitHub.
>

tschw commented:
The `.needUpdate

bhouston commented:
@mattdesl wrote:
> My point was, even if you set map.needsUpdate to true, it will not recompile the shader with new defines.

Generally people will not change a texture's encoding on the fly, as only one encoding per texture is really valid.  So I think we do not really need live updating for this property.

bhouston commented:
BTW @mattdesl @tschw thanks a lot for your attention on this PR, I greatly appreciate it (and thanks to and @mrdoob @WestLangley on the other PRs.)   It is collaborations like this that make contributing to ThreeJS rewarding. :)

bhouston committed:
add better documentation that MAterial.needsUpdate needs to be set if one changes Texture.encoding.

bhouston commented:
In regards to @mattdesl's concern that only support maps support encoding, expressed here: https://github.com/mrdoob/three.js/pull/8177#issuecomment-186859429  What is the consensus?  To get this PR accepted, should I add encoding support to roughnessMap, specularMap, lightMap, aoMap, metalnessMap, normalMap, bumpMap, displacementMap?  If the consensus is that we should do this because it is only costs a bit more in terms of compile time, I can do that.  My opinion, which is only an weak option, is that generally it shouldn't be necessary but I am willing to go with the majority option on this if it is in the other direction.

bhouston committed:
add comment in docs.

bhouston commented:
@mattdesl @tschw I've added documentation to the THREE.Texture constructor as well as the formal THREE.Texture documentation file stating that one needs to trigger a Material.needsUpdate in order to support dynamically changing encodings.

I hope this is pretty close to sufficient for getting this PR accepted?  I really do not want to take on re-doing the ThreeJS inter-object updating scheme in this PR, it is big task.

mattdesl commented:
I think it's good to merge.

After merge I can go through and add back the native sRGB stuff for WebGL2. It will introduce another workflow issue: let's say we set sRGB on a texture, but we have to fallback to faking it in a shader for whatever reason (e.g. it's a compressed or float type texture). If you decide to re-write the phong/whatever material with RawShaderMaterial, you have no way of querying which encoding is *actually* going to be used on a texture. We can explore all that in another issue, though.

tschw commented:
@mattdesl 
> If you decide to re-write the phong/whatever material with RawShaderMaterial, you have no way of querying which encoding is actually going to be used on a texture. We can explore all that in another issue, though.

Why would you want to do that? Isn't `ShaderMaterial` the appropriate wrapper for the custom shader in that case? And then the information should be in `.defines`, which is good because it doesn't eat up precious uniform slots in the fragment shader...

mattdesl commented:
I use raw often for various reasons.

- It plays nicely with my preprocessor ([glslify](https://github.com/stackgl/glslify))
- It ensures I won't end up with naming conflicts and wasted attribute/uniform/etc slots
- It is the only way to use `#version 300 es` right now in ThreeJS

But yeah, we can discuss that in more detail in another issue. :smile:

tschw commented:
@mattdesl 
> [... `RawShaderMaterial` ...]
> It ensures I won't end up with naming conflicts and wasted attribute/uniform/etc slots

Oh well, if you want state it's gotta be named somehow. But only what gets "statically referenced" (it's the term used in the spec) in your code actually counts against the attribute/uniform/etc limit.

> But yeah, we can discuss that in more detail in another issue.

My point was, that you shouldn't be too surprised that information is not around and has to be passed in manually when explicitly asking for it :-).

mrdoob commented on src/renderers/shaders/ShaderChunk/emissivemap_pars_fragment.glsl:
```diff
@@ -2,4 +2,10 @@
 
+	vec4 emissiveMapTexelToLinear( vec4 value ) {
+		#define MACRO_DECODE EMISSIVEMAP_ENCODING
+			#include <encoding_template>
```
I think we should move this logic to JavaScript, somehow...

bhouston replied:
If we had a better preprocessor I believe we could.  I really struggled with the limitations of glsl in this case to avoid using up precious uniforms for the encoding selection.

BTW with this new preprocessor that supports includes we can gut ShaderLib. :)  What I suggest is we refactor the large single-level WebGL/shaders/chunks directory so that we have a directory for each shader (with a frag/vert shader each) and then a directory (at least) of shared chunks.

Both of these above changes can be done after this is merged though. :)  Also I believe that @tschw is the compiler expert, although I am not suggesting he has time to look into this -- I sort of fake being a compiler guy.

bhouston replied:
I guess one approach to moving this to JavaScript more would be to just define make "mapTexelDecode" a define which resolves into a macro that calls the decoder function we are looking to use.  I think that could work.  Removes the need for the "include" system though....

mrdoob replied:
> I guess one approach to moving this to JavaScript more would be to just define make "mapTexelDecode" a define which resolves into a macro that calls the decoder function we are looking to use. I think that could work. 

Nice! I like this approach better 😁

bhouston committed:
start addressing @mrdoob's feedback here: https://github.com/mrdoob/three.js/pull/8117/files#r53900857

bhouston committed:
Merge branch 'texture_encoding' of git@github.com:bhouston/three.js.git into texture_encoding

bhouston committed:
remove encoding_template and WebGLShaderPrePRocessor as mrdoob has rewritten it.

bhouston committed:
Merge branch 'dev' of git@github.com:mrdoob/three.js.git into texture_encoding

# Conflicts:
#	src/renderers/shaders/ShaderLib.js

bhouston committed:
add encodings to includes for primary frag shaders.

bhouston committed:
remove envMap encoding macro.

bhouston committed:
remove remnants of WebGLShaderPreProcessor, re-add mistakenly removed line.

bhouston commented:
@mrdoob I have updated this PR with two changes in response to your feedback:

(1) I've replaced the include-based function definition with JavaScript-based function generation here, per your request:

https://github.com/mrdoob/three.js/pull/8117/files#diff-58e8247d21441c70f4f66e068eac5bfaR10

(2) I've merged in your dev branch that has the new regex-based include system, and I have removed the more verbose include-system WebGLShaderPreprocessor.

mrdoob commented:
Great work!

mrdoob commented:
Thanks!

bhouston commented:
Thanks for being open to the change!  Now onto the PMREM!

mrdoob commented:
Seems like it produced some breakage though...

`webgl_animation_cloth` is way more bright now.
`webgl_effects_cardboard` fails to compile.
...

bhouston commented:
I'll have a look now.

BTW there will be some minor look changes because texture decoding no longer pays attention to WebGLRenderer.inputGamma = true/false, instead you have to set each texture.encoding = THREE.GammaEncoding to have similar gamma result.

mrdoob commented:
I see... I'm thinking that it's probably better to break backwards compatibility and show a warning when using the user sets/access `gammaInput`. We still use `gammaFactor` and `gammaOutput`, right?

mrdoob commented:
Also, out of curiosity... how come we're not using [gl.UNPACK_COLORSPACE_CONVERSION_WEBGL](https://msdn.microsoft.com/en-us/library/dn302429(v=vs.85).aspx)?

bhouston commented:
In the other PR I added perfect backwards compatibility with inputGamma.
That is how I got the grass in the cloth scene darker.

Best regards,
Ben Houston
http://Clara.io Online 3d modeling and rendering
On Feb 25, 2016 1:44 AM, "Mr.doob" <notifications@github.com> wrote:

> I see... I'm thinking that it's probably better to break backwards
> compatibility and show a warning when using the user sets/access
> gammaInput. We still use gammaFactor and gammaOutput, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mrdoob/three.js/pull/8117#issuecomment-188638272>.
>
